File: jas/geometry/regular_shapes.py
"""Shape-geometry helpers for regular polygons and stars.

Python analogue of jas_dioxus/src/geometry/regular_shapes.rs,
JasSwift/Sources/Geometry/RegularShapes.swift, and
jas_ocaml/lib/geometry/regular_shapes.ml.

L2 primitives per NATIVE_BOUNDARY.md §5 — shape geometry is shared
across vector-illustration apps.
"""

from __future__ import annotations

import math

# Ratio of inner radius to outer radius for the default star.
STAR_INNER_RATIO: float = 0.4


def regular_polygon_points(
    x1: float, y1: float, x2: float, y2: float, n: int,
) -> list[tuple[float, float]]:
    """Compute vertices of a regular N-gon whose first edge runs from
    (x1, y1) to (x2, y2). Returns ``n`` (x, y) pairs. For degenerate
    zero-length edges returns ``n`` copies of the start point.
    """
    ex = x2 - x1
    ey = y2 - y1
    s = math.hypot(ex, ey)
    if s == 0.0:
        return [(x1, y1)] * n
    mx = (x1 + x2) / 2.0
    my = (y1 + y2) / 2.0
    px = -ey / s
    py = ex / s
    d = s / (2.0 * math.tan(math.pi / n))
    cx = mx + d * px
    cy = my + d * py
    r = s / (2.0 * math.sin(math.pi / n))
    theta0 = math.atan2(y1 - cy, x1 - cx)
    return [
        (cx + r * math.cos(theta0 + 2.0 * math.pi * k / n),
         cy + r * math.sin(theta0 + 2.0 * math.pi * k / n))
        for k in range(n)
    ]


def star_points(
    sx: float, sy: float, ex: float, ey: float, points: int,
) -> list[tuple[float, float]]:
    """Compute vertices of a star inscribed in the axis-aligned
    bounding box with corners (sx, sy) and (ex, ey). ``points`` is
    the number of outer vertices; the returned list alternates
    outer/inner for ``2 * points`` total. The first outer point
    sits at top-center."""
    cx = (sx + ex) / 2.0
    cy = (sy + ey) / 2.0
    rx_outer = abs(ex - sx) / 2.0
    ry_outer = abs(ey - sy) / 2.0
    rx_inner = rx_outer * STAR_INNER_RATIO
    ry_inner = ry_outer * STAR_INNER_RATIO
    n = points * 2
    theta0 = -math.pi / 2.0
    out: list[tuple[float, float]] = []
    for k in range(n):
        angle = theta0 + math.pi * k / points
        rx, ry = (rx_outer, ry_outer) if k % 2 == 0 else (rx_inner, ry_inner)
        out.append((cx + rx * math.cos(angle), cy + ry * math.sin(angle)))
    return out

File: jas_dioxus/src/geometry/regular_shapes.rs
//! Shape-geometry helpers for regular polygons and stars.
//!
//! These functions were originally inlined in `tools/polygon_tool.rs`
//! and `tools/star_tool.rs`. When those native tools were migrated
//! to YAML-driven YamlTool (see RUST_TOOL_RUNTIME.md), the geometry
//! moved here so it survives the deletion and stays available both
//! to the `polygon` / `star` arms of `interpreter::effects::build_element`
//! and to the matching overlay render types in `tools::yaml_tool`.
//!
//! Per NATIVE_BOUNDARY.md §5 ("Domain (L2) primitives"), shape
//! geometry is legitimately native — it's a universal vector-app
//! primitive, not app-specific behavior.

use std::f64::consts::PI;

/// Ratio of inner radius to outer radius for the default star.
pub const STAR_INNER_RATIO: f64 = 0.4;

/// Compute vertices of a regular N-gon whose first edge runs from
/// `(x1, y1)` to `(x2, y2)`. Returns a `Vec` of `n` `(x, y)` pairs.
///
/// For `n < 3`, the result is geometrically degenerate but still
/// well-defined (the formula still returns `n` points around a
/// computed center). Callers should validate `n >= 3` upstream.
pub fn regular_polygon_points(
    x1: f64,
    y1: f64,
    x2: f64,
    y2: f64,
    n: usize,
) -> Vec<(f64, f64)> {
    let ex = x2 - x1;
    let ey = y2 - y1;
    let s = (ex * ex + ey * ey).sqrt();
    if s == 0.0 {
        return vec![(x1, y1); n];
    }
    let mx = (x1 + x2) / 2.0;
    let my = (y1 + y2) / 2.0;
    let px = -ey / s;
    let py = ex / s;
    let d = s / (2.0 * (PI / n as f64).tan());
    let cx = mx + d * px;
    let cy = my + d * py;
    let r = s / (2.0 * (PI / n as f64).sin());
    let theta0 = (y1 - cy).atan2(x1 - cx);
    (0..n)
        .map(|k| {
            let angle = theta0 + 2.0 * PI * k as f64 / n as f64;
            (cx + r * angle.cos(), cy + r * angle.sin())
        })
        .collect()
}

/// Compute vertices of a star inscribed in the axis-aligned bounding
/// box with corners `(sx, sy)` and `(ex, ey)`. `points` is the number
/// of outer vertices; the returned `Vec` alternates outer / inner
/// points for `2 * points` total. Outer radius matches the bounding
/// box; inner radius is `STAR_INNER_RATIO × outer`. First outer point
/// sits at the top-center of the bounding box.
pub fn star_points(
    sx: f64,
    sy: f64,
    ex: f64,
    ey: f64,
    points: usize,
) -> Vec<(f64, f64)> {
    let cx = (sx + ex) / 2.0;
    let cy = (sy + ey) / 2.0;
    let rx_outer = (ex - sx).abs() / 2.0;
    let ry_outer = (ey - sy).abs() / 2.0;
    let rx_inner = rx_outer * STAR_INNER_RATIO;
    let ry_inner = ry_outer * STAR_INNER_RATIO;
    let n = points * 2;
    let theta0 = -PI / 2.0;
    (0..n)
        .map(|k| {
            let angle = theta0 + PI * k as f64 / points as f64;
            let (rx, ry) = if k % 2 == 0 {
                (rx_outer, ry_outer)
            } else {
                (rx_inner, ry_inner)
            };
            (cx + rx * angle.cos(), cy + ry * angle.sin())
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn star_points_count_is_double_outer() {
        let pts = star_points(0.0, 0.0, 100.0, 100.0, 5);
        assert_eq!(pts.len(), 10);
    }

    #[test]
    fn star_points_first_at_top() {
        let pts = star_points(0.0, 0.0, 100.0, 100.0, 5);
        assert!((pts[0].0 - 50.0).abs() < 1e-9);
        assert!((pts[0].1 - 0.0).abs() < 1e-9);
    }

    #[test]
    fn star_points_alternate_inner_outer() {
        let pts = star_points(0.0, 0.0, 100.0, 100.0, 5);
        let center = (50.0, 50.0);
        for (i, (x, y)) in pts.iter().enumerate() {
            let dx = x - center.0;
            let dy = y - center.1;
            let r = (dx * dx + dy * dy).sqrt();
            let expected = if i % 2 == 0 { 50.0 } else { 20.0 };
            assert!(
                (r - expected).abs() < 1e-9,
                "point {} radius {} != {}",
                i, r, expected,
            );
        }
    }

    #[test]
    fn regular_polygon_first_edge_matches_input() {
        let pts = regular_polygon_points(0.0, 0.0, 10.0, 0.0, 5);
        assert_eq!(pts.len(), 5);
        // First vertex is (x1, y1) up to floating-point noise.
        assert!((pts[0].0 - 0.0).abs() < 1e-9);
        assert!((pts[0].1 - 0.0).abs() < 1e-9);
        // Second vertex is (x2, y2) — the other end of the first edge.
        assert!((pts[1].0 - 10.0).abs() < 1e-9);
        assert!((pts[1].1 - 0.0).abs() < 1e-9);
    }

    #[test]
    fn regular_polygon_zero_length_edge_returns_repeated_point() {
        // Degenerate input: both endpoints coincide.
        let pts = regular_polygon_points(5.0, 7.0, 5.0, 7.0, 5);
        assert_eq!(pts.len(), 5);
        for p in &pts {
            assert_eq!(*p, (5.0, 7.0));
        }
    }
}
